### PR TITLE
[CARBONDATA-3065]make inverted index false by default

### PR DESCRIPTION
**Pull request by akashrn5**
### Bottleneck with invertedIndex:
1. As for each page first we will sort the data and generate inverted index, data loading performance will get impacted.because of this
2. Store size is more because of stroing inverted index for each dimension column which results in more IO and it impacts query performance 
3. One extra lookup happenes during query due to presence  of inverted index which is causing many cachline miss and it impacts the query performance

### Changes Proposed in this PR:
1. by default inverted index will be disabled for all the dimension columns, only row level sorting happens in sortstep and column level sorting will be skipped.
2. new table property will be added called `INVERTED_INDEX` which takes comma separated column names as value, and only for these columns inverted index will be generated.
3. `NO_INVERTED_INDEX` table property will be deprecated, and it will be kept only for compatibility.
4. Handled in describe formatted for inverted index columns

### how this tested
existing UTs will take care and new test cases are added.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1180/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1393/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9444/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1403/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1189/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1407/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1195/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9456/

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 1]
```diff
@@ -359,8 +359,13 @@ private CarbonCommonConstants() {
   public static final String TABLE_BLOCKSIZE = "table_blocksize";
   // table blocklet size in MB
   public static final String TABLE_BLOCKLET_SIZE = "table_blocklet_size";
-  // set in column level to disable inverted index
+  /**
+   * set in column level to disable inverted index
+   * @Deprecated :This property is deprecated, it is kep just for compatibility
```
Comment (kevinjmh) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: spelling: kept

**Comment by kevinjmh** [comment]
The InvertedIndex/NoInvertedIndex setting is confusing.
1. the value `isInvertedIndex` assigned to different IndexCodec  in `createEncoderForDimensionLegacy` requires us to set the column both SortColumns and use InvertedIndex. What if I set it in INVERTED_INDEX but not in SORT_COLUMNS？
2. what the boolean value `isInvertedIndex` in IndexCodec do is to control whether to do RLE on datapage?

These make the setting not a direct switch to control how the data proceed

**Comment by akashrn5** [comment]
@kevinjmh 
1. when you set in INVERTED_INDEX , but not in SORT_COLUMNS, then data will not be sorted, only RLE will be applied on data.
2. `isInvertedIndex ` basically this boolean cannot say that, it is do RLE, basically RLE will be applied on both inverted and no inverted case, but after checking `isInvertedIndex ` it decides to sort based on isSort check

**Comment by kevinjmh** [comment]
@akashrn5 thanks for reply. 

1. Let's take a detail case. you can check whether it is right.
 
 In `DictDimensionIndexCodec#createEncoder`, as the setting I said above
 `isSort`=false
 `isDoInvertedIndex` = true
 `isInvertedIndex`=`isSort`&&`isDoInvertedIndex` = false
so, it will go to `indexStorage = new BlockIndexerStorageForNoInvertedIndexForShort(data, false);`.
In the construction method, we can see that it only assigns the dataPage value. No RLE.
```
  public BlockIndexerStorageForNoInvertedIndexForShort(byte[][] dataPage, boolean applyRLE) {
    this.dataPage = dataPage;
    if (applyRLE) {
      List<byte[]> actualDataList = new ArrayList<>();
      for (int i = 0; i < dataPage.length; i++) {
        actualDataList.add(dataPage[i]);
      }
      rleEncodeOnData(actualDataList);
    }
  }
```

2. If isInvertedIndex is TRUE, then the isSort check must be TRUE

```
isInvertedIndex    =      isSort    &&        isDoInvertedIndex;
       ^                     ^                       ^
       |                     |                       |
internalUsed          SORT_COLUMNS           INVERTED_INDEX

```

**Comment by akashrn5** [comment]
> @akashrn5 Please expose these properties from SDK and fileformat as well.

@ravipesala handed for SDK and fileformat, please review

**Comment by akashrn5** [comment]
@kevinjmh basically, with my changes the column level sorting will be skipped and only row level sorting will be done in sort step.

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java (hunk 1 above)
Comment (akashrn5) [reply core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1238/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1240/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1241/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1457/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9506/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1245/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1461/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9510/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1247/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1463/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9512/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/TestNoInvertedIndexLoadAndQuery.scala [hunk 2]
```diff
@@ -289,6 +292,44 @@ class TestNoInvertedIndexLoadAndQuery extends QueryTest with BeforeAndAfterAll {
...
+    sql(
+      s"""
+           LOAD DATA LOCAL INPATH '$testData1' into table index1
+           """)
+    checkAnswer(
+      sql(
+        """
+           SELECT * FROM index1 WHERE city = "Bangalore"
+        """),
+      Seq(Row(19.0, "Emily", "Bangalore")))
```
Comment (ravipesala) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/TestNoInvertedIndexLoadAndQuery.scala]: Try to get the CarbonTable using CarbonMetadata.getInstance and verify the inverted index columns for all the testcases

**Comment by ravipesala** [comment]
@akashrn5 In existing flow inverted index is applied by default for all sort columns . No sort columns and measure columns cannot be applied for an inverted index. Is it the same case now also?  Are we allowing inverted index only for sort columns or allowing for all columns?  If allowed only for sort columns then please add the validations in parser and also in sdk.

**Comment by akashrn5** [comment]
@ravipesala
in existing flow, all the sort columns and dimensions will have inverted index and,
 in current implementation, if measure columns are given in inverted_index property, then i ll not set it as inverted index. Basically i wont throw any exception. Do we need to throw exception?
and currently im allowing the inverted index for all dimension columns and not restricting to sort columns only. 
So basically, user can mention columns in inverted_index columns (dimensions), i am setting those as inverted index. it is correct right

**Comment by ravipesala** [comment]
ok, its fine

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/TestNoInvertedIndexLoadAndQuery.scala (hunk 2 above)
Comment (akashrn5) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/TestNoInvertedIndexLoadAndQuery.scala]: done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1269/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1484/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9533/

**Comment by akashrn5** [comment]
2.2.1 and 2.3.1 failed test cases, which are not related and these are random

**Comment by akashrn5** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1494/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1282/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9543/

**Comment by ravipesala** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1359/

**Comment by akashrn5** [comment]
@ravipesala please review the documentation commit added

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1360/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1570/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9618/

**Comment by akashrn5** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1374/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1584/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9632/

**Comment by ravipesala** [comment]
LGTM
